File: solutions/84/main.py
from typing import List 

class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        result = 0

        heights.insert(0, 0)
        heights.append(0)

        stack = []
        stack.append(0)
        for i in range(1, len(heights)):
            if heights[stack[-1]] < heights[i]:
                stack.append(i)
            elif heights[stack[-1]] == heights[i]:
                stack[-1] = i
            else:
                while len(stack) > 0 and heights[stack[-1]] > heights[i]:
                    mid = heights[stack[-1]]
                    stack.pop()
                    if len(stack) > 0:
                        left = stack[-1]
                        w = i - left - 1
                        result = max(result, w * mid)
            stack.append(i)
        return result

if __name__ == "__main__": 
    heights = [2,1,5,6,2,3]
    s = Solution()
    assert s.largestRectangleArea(heights) == 10

File: solutions/84/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Sat Nov 27 21:36:13 2021
> Description:   
 ************************************************************************/
#include <vector> 
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        vector<int> s;
        heights.push_back(0);
        int sum = 0;
        int i = 0;
        while (i < heights.size()) {
            if (s.empty() || heights[i] > heights[s.back()]) {
                s.push_back(i);
                ++i;
            }
            else
            {
                int t = s.back();
                s.pop_back();
                // when s is empty, width must be i, since all previous heights must be smaller than heights[t] 
                sum = max(sum, heights[t] * (s.empty() ? i : i - s.back() - 1));
            }
        }
        return sum;
    }
};

TEST(Test84, SimpleTest) {
    vector<int> heights{
        2,1,5,6,2,3
    };
    Solution s; 
    EXPECT_EQ(s.largestRectangleArea(heights), 10);
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}
